### PR TITLE
ApiControllersWithContext

### PR DESCRIPTION
**Pull request by prafullbhosale**

### CODE REVIEW

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+using SampleWebApp.Data;
+using SampleWebApp.Models;
+
+namespace SampleWebApp.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/AuthorssApi")]
+    public class AuthorssApiController : Controller
+    {
+        private AuthorDataContext _context;
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:17]: Readonly?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        public IActionResult GetAuthor([FromRoute] int id)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
+
+            if (author == null)
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:42]: Consider conditional operator?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            {
+                return BadRequest(ModelState);
+            }
+
+            if (id != author.AuthorID)
+            {
+                return BadRequest();
+            }
+
+            _context.Entry(author).State = EntityState.Modified;
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:64]: Why not using the Update method?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+            try
+            {
+                _context.SaveChanges();
+            }
+            catch (DbUpdateConcurrencyException)
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:70]: Catching concurrency exception here, but not in the other pull request.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            if (disposing)
+            {
+                _context.Dispose();
+            }
+            base.Dispose(disposing);
+        }
+
+        private bool AuthorExists(int id)
+        {
+            return _context.Author.Count(e => e.AuthorID == id) > 0;
```
Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:146]: Use Any.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using SampleWebApp.Data;
+using SampleWebApp.Models;
+
+namespace SampleWebApp.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/AuthorssApi")]
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:14]: Was `AuthorssApi` the the name that was input by the user in this case? You wouldn't normally double-down on "api" in a route.

The `api` prefix is a legacy webapi-ism and isn't required. (Some kind of routing prefix used be be required to use both WebAPI and MVC in the same site). It's not wrong, it's just an old convention based on old requirements. Not sure if others have an opinion.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
+                }
+            }
+
+            return new StatusCodeResult(StatusCodes.Status204NoContent);
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:82]: `return NoContent();`

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                {
+                    return new StatusCodeResult(StatusCodes.Status409Conflict);
+                }
+                else
+                {
+                    throw;
+                }
+            }
+
+            return CreatedAtRoute("GetAuthor", new { id = author.AuthorID }, author);
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:111]: `CreatedAtAction("GetAuthor", ...); would work here now, this was a limitation in legacy WebAPI which means that the name for the route on`GetAuthor` can also be removed.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            return Ok(author);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _context.Dispose();
+            }
+            base.Dispose(disposing);
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:141]: This shouldn't be needed (and isn't really correct IMO). The DbContext comes from DI and the DI will dispose it

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using SampleWebApp.Data;
+using SampleWebApp.Models;
+
+namespace SampleWebApp.Controllers
+{
+    [Produces("application/json")]
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:13]: What's the purpose of putting produces here? Just metadata for swagger? If that's the case we need to put some metadata on the actions returning `IActionResult`

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        // GET: api/AuthorssApi/5
+        [HttpGet("{id}", Name = "GetAuthor")]
+        public IActionResult GetAuthor([FromRoute] int id)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            Author author = _context.Author.SingleOrDefault(m => m.AuthorID == id);
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:40]: Was the `.Author` property generated as part of this scaffold? Is that preferable to `Authors`?

File: src/SampleWebApp/Startup.cs [hunk 12]
```diff
@@ -41,6 +42,11 @@ public Startup(IHostingEnvironment env, IApplicationEnvironment appEnv)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddDbContext<BookDataContext>(options =>
+                    options.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=BookDataContext-36f2c634-aa7b-4dbd-80e6-95e06e3521a3;Trusted_Connection=True;MultipleActiveResultSets=true"));
```
Comment (rynowak) [inline src/SampleWebApp/Startup.cs:46]: Wait - did scaffolding generate two different databases for this? 

We would definitely not recommend putting the connection strings for these in code right?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            if (author == null)
+            {
+                return NotFound();
+            }
+
+            _context.Author.Remove(author);
```
Comment (rynowak) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:129]: @ajcvickers should we be using the async variants of these things?

I know in legacy WebAPI/MVC we had the option to do async or not... is that really relevant 3 years later?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 6 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:14]: Yes, `AuthorssApiController` was user input for the controller name. The suggested name on the UI would be AuthorsController.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 11 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:40]: Yes, the `Author` property was generated as part of the scaffold. EF7 doesn't support pluralization which was being originally used to get the pluralized names for the DbSet properties. 

We had discussions on getting a dumbed down approach to indicate that this is a collection, (like adding 'DbSet' suffix to the property name) but never got to agreement.

File: src/SampleWebApp/Startup.cs (hunk 12 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Startup.cs:46]: In this case yes, the AuthorssApiController and BooksApiController were generated as part of two separate scaffolds. In each I provided a new DataContext. If user input provides an existing DataContext, scaffolding will just try to add the model to the existing one. 

We do have an open issue for moving the connection string to appsettings.json, but won't be targeted for RC2.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 13 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:129]: The async / sync is based on user input. The scaffolding dialog has a checkbox which passes in a flag for this. 

In this case I happened to not pass in that flag.

Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:129]: Yes. I believe @divega commented that somewhere else already.

Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:129]: For the queries and SaveChanges, that is.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 10 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:13]: This seems to be from the pre RC1 days. 
The only thing I could find out is that this was a way for content negotiation. 
Am not sure if this is needed or not. 

@rynowak do you know who would be able to help out with this?

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 9 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:141]: Will remove this method.

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 5 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:146]: Will change to `return _context.Author.Where(e => e.AuthorID == id).Any();`

Comment (ajcvickers) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:146]: Why not this? `_context.Author.Any(e => e.AuthorID == id);`

Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:146]: yeah that makes better sense. 
Will change to `_context.Author.Any(e => e.AuthorID == id);`

File: src/SampleWebApp/Controllers/AuthorssApiController.cs (hunk 8 above)
Comment (prafullbhosale) [inline src/SampleWebApp/Controllers/AuthorssApiController.cs:111]: Ok, will change to `return CreatedAtAction("GetAuthor", new { id = author.AuthorID }, author);`

Will also change 

```
[HttpGet("{id}", Name = "GetAuthor")]
public IActionResult GetAuthor([FromRoute] int id)
 {
```

to 

```
[HttpGet("{id}")]
public IActionResult GetAuthor([FromRoute] int id)
{
```

**Pull request closed without merging**